### PR TITLE
Fix to network desyncs related to player movement

### PR DESCRIPTION
**Pull request by FluffyQuack**
This fixes the first desync issue listed here: https://github.com/diasurgical/devilutionX/issues/1576

Video example of the issue: https://www.youtube.com/watch?v=wDXCDKLnpNY

The cause of the problem is the _ptarx and _ptary values (these correspond to where the player is pathfinding to) are sometimes set incorrectly. The game keeps sending these values (and player position) to other clients so they can ensure the player position is the same across all clients, but sometimes the values are set incorrectly which makes the local player tell other clients to move the player to the wrong position.

For instance, this desync is caused by the following sequence of code:

- Player is in the middle of the attack animation.
- Player clicks to move, which makes the game pathfind to a new location, and it sets _ptarx and _ptary to where the player is pathfinding to.
- Game waits for player to finish attacking before starting to move.
- Player finishes the attack animation, this calls PlayerStand() which then calls FixPlayerLocation() which resets _ptarx and _ptary to the player's current position.
- Player starts sending the new (wrong) _ptarx and _ptary to other clients, and other clients keep pathfinding the player to the wrong location.

My fix re-calculates the _ptarx and _ptary values each time the player starts walking towards a new tile when following pathfinding. Based on brief testing, this fixes the issue entirely and I haven't seen any downsides.

A better fix would probably be to prevent the game from resetting _ptarx and _ptary to the wrong value after finishing an attack.

### CODE REVIEW

**Comment by FluffyQuack** [comment]
I've got a better idea for how to implement this fix, and I've changed this PR to be an early version of that.

ptargx/ptargy are only ever used for 2 things in the code:

- They keep being sent to other players so they can fix the position of our local player.
- It's used for in the "holding left mouse button to move" code to check if we're highlighting a different tile than the one we're moving to.

Since the game does such a bad job correctly updating those two variables (and that's why it's so easy to cause desyncs with player movement), why not just calculate the values from scratch based on the player's pathfinding each time they're used? That also means we can delete a lot of code where the game tries to set them to various values based on player states, which would make it easier to overhaul code related to this in the long run.

I recorded a video showcasing how easily you can break network sync with basic player movement (I do 3 things in the video: queue a move while attacking; cancel movement with attacking and then moving again; and simply clicking to move back and forth), and how it looks with this fix applied: https://www.youtube.com/watch?v=I1ONHhJ2xg0

**Review by AJenbo — DISMISSED** [review]

File: Source/multi.cpp [hunk 1]
```diff
@@ -114,6 +114,25 @@ static BYTE *multi_recv_packet(TBuffer *pBuf, BYTE *body, DWORD *size)
 
 static void NetRecvPlrData(TPkt *pkt)
 {
+	/*
+	* We calculate _ptargx and _ptargy here and prepare for it to be sent to other clients.
+	* If px/py and ptarx/ptary don't match on other clients, then they'll move our local player to that location.
+	*/
+	{
+		// clang-format off
+		int directionOffsetX[8] = {  0,-1, 1, 0,-1, 1, 1,-1 };
```
Comment (AJenbo) [inline Source/multi.cpp:123]: Would be nice if you could make your editor apply the clang-format rules.

Comment (FluffyQuack) [reply Source/multi.cpp:123]: I turned it off there because the arrays became hard to read when you can't see the XY values side-by-side.

**Comment by FluffyQuack** [comment]
Okay, I've done some changes and I think this is the best implementation for this fix.

I've simplified how the _ptargx/_ptargy values are used. Besides initialization, they're now only being updated in one place, which makes it easier to understand how the values work.

The values were also used for input (checking if mouse is targeting a tile player is already moving to), but it was enough to replace that check.

I like this implementation because I think it's easy to work from. If anyone wants to actually try implementing a netcode which is sync thanks to fully deterministic gameplay simulation, then the _ptargx/_ptargy values could just be removed entirely. Or someone could try to find ways to correct desyncs (that would be easier to implement, but it wouldn't fully fix the desyncs).

**Comment by AJenbo** [comment]
Have you considered simply not allowing playermovment?

**Comment by FluffyQuack** [comment]
> Have you considered simply not allowing playermovment?

Could fix all desyncs by setting max player count to 1.

File: Source/player.h [hunk 2]
```diff
@@ -160,8 +160,14 @@ struct PlayerStruct {
...
 	Sint32 _pfuty;   // Future tile Y-position of player. Set at start of walking animation
-	Sint32 _ptargx;  // Target tile X-position for player movment. Set during pathfinding
-	Sint32 _ptargy;  // Target tile Y-position for player movment. Set during pathfinding
+
+	/*
+	* These values store the target position for non-local players during network play.
+	* This is needed because player positions drift over time due to desyncs, so we have
+	* clients try to pathfind non-local players to this position.
+	*/
+	Sint32 _ptargx, _ptargy;
```
Comment (AJenbo) [inline Source/player.h:169]: I don't think this is going to jive with clang-tidy, though it makes good sense. Maybe we should instead use a point type, that would greatly simplify all the structures as well.

File: Source/multi.cpp [hunk 3]
```diff
@@ -114,11 +114,14 @@ static BYTE *multi_recv_packet(TBuffer *pBuf, BYTE *body, DWORD *size)
 
 static void NetRecvPlrData(TPkt *pkt)
 {
+	Sint32 targetX, targetY;
```
Comment (AJenbo) [inline Source/multi.cpp:117]: Oh the irony, I just finished a PR that changes this to `int`.

File: Source/player.cpp [hunk 4]
```diff
@@ -3131,6 +3129,24 @@ bool PM_DoNewLvl(int pnum)
 	return false;
 }
 
+void GetPlayerTargetPosition(int playerNumber, Sint32 *x, Sint32 *y)
```
Comment (AJenbo) [inline Source/player.cpp:3132]:
```suggestion
void UpdatePlayerTargetPosition(PlayerStruct &player)
```

File: Source/player.h (hunk 2 above)
Comment (FluffyQuack) [reply Source/player.h:169]: Should I just change it back to this?

```
Sint32 _ptargx;
Sint32 _ptargy;
```

Comment (AJenbo) [reply Source/player.h:169]: No, lets try it out and see what happens once I get to this file :)

**Comment by AJenbo** [comment]
@FluffyQuack I merged my clean up and rebased your code on top of it, so merged manually via this commit: https://github.com/diasurgical/devilutionX/commit/4f417d2fdcb2ed32f55d51bc25b5525b9c74f888

I made it into a member function and used a range so we don't even have to bother with what types any of this is, hope you like the new style :) And feel free to ask questions if you don't get some of it.

**Pull request closed without merging**